std-bits/table/src/main/java/org/enso/table/data/column/builder/InferredBuilder.java

```diff
@@ -155,7 +155,13 @@ private void retypeAndAppend(Object o) {
   }
 
   private void retypeToMixed() {
-    ObjectBuilder objectBuilder = new MixedBuilder(initialSize);
+    // The new internal builder must be at least `currentSize` so it can store
+    // all the current values. It must also be at least 'initialSize' since the
+    // caller might be using appendNoGrow and is expecting to write at least
+    // that many values.
+    int capacity = Math.max(initialSize, currentSize);
```

Comment:
https://github.com/enso-org/enso/issues/8030 -- investigate if this happens elsewhere.